src/iceberg/expression/inclusive_metrics_evaluator.cc

```diff
@@ -0,0 +1,546 @@
+        reference != nullptr) {
+      return ParseLowerBound(*reference);
+    } else if (auto transform = std::dynamic_pointer_cast<BoundTransform>(expr);
+               transform != nullptr) {
+      return TransformLowerBound(*transform);
+    } else {
+      return std::nullopt;
+    }
+    // TODO(xiao.dong) handle extract lower and upper bounds
+  }
+
+  Result<std::optional<Literal>> UpperBound(const std::shared_ptr<Bound>& expr) {
+    if (auto reference = std::dynamic_pointer_cast<BoundReference>(expr);
+        reference != nullptr) {
+      return ParseUpperBound(*reference);
+    } else if (auto transform = std::dynamic_pointer_cast<BoundTransform>(expr);
+               transform != nullptr) {
+      return TransformUpperBound(*transform);
+    } else {
+      return std::nullopt;
+    }
+    // TODO(xiao.dong) handle extract lower and upper bounds
+  }
+
+  Result<std::optional<Literal>> ParseLowerBound(const BoundReference& ref) {
+    int32_t id = ref.field().field_id();
+    auto type = ref.type();
+    if (!type->is_primitive()) {
+      return NotSupported("Lower bound of non-primitive type is not supported.");
+    }
+    auto primitive_type = internal::checked_pointer_cast<PrimitiveType>(type);
+    if (!data_file_.lower_bounds.empty() && data_file_.lower_bounds.contains(id)) {
+      ICEBERG_ASSIGN_OR_RAISE(
+          auto lower,
+          Literal::Deserialize(data_file_.lower_bounds.at(id), primitive_type));
+      return lower;
+    }
+
+    return std::nullopt;
+  }
+
+  Result<std::optional<Literal>> ParseUpperBound(const BoundReference& ref) {
+    int32_t id = ref.field().field_id();
+    auto type = ref.type();
+    if (!type->is_primitive()) {
+      return NotSupported("Upper bound of non-primitive type is not supported.");
+    }
+    auto primitive_type = internal::checked_pointer_cast<PrimitiveType>(type);
+    if (!data_file_.upper_bounds.empty() && data_file_.upper_bounds.contains(id)) {
+      ICEBERG_ASSIGN_OR_RAISE(
+          auto upper,
+          Literal::Deserialize(data_file_.upper_bounds.at(id), primitive_type));
+      return upper;
+    }
+
+    return std::nullopt;
+  }
+
+  Result<std::optional<Literal>> TransformLowerBound(BoundTransform& boundTransform) {
+    auto transform = boundTransform.transform();
+    if (transform->PreservesOrder()) {
+      ICEBERG_ASSIGN_OR_RAISE(auto lower, ParseLowerBound(*boundTransform.reference()));
+      ICEBERG_ASSIGN_OR_RAISE(auto transform_func,
+                              transform->Bind(boundTransform.reference()->type()));
+      if (lower.has_value()) {
+        return transform_func->Transform(lower.value());
+      }
+    }
+
+    return std::nullopt;
+  }
+
+  Result<std::optional<Literal>> TransformUpperBound(BoundTransform& boundTransform) {
+    auto transform = boundTransform.transform();
+    if (transform->PreservesOrder()) {
+      ICEBERG_ASSIGN_OR_RAISE(auto upper, ParseLowerBound(*boundTransform.reference()));
```

Comment:
```suggestion
      ICEBERG_ASSIGN_OR_RAISE(auto upper, ParseUpperBound(*boundTransform.reference()));
```

We might need to port the TestInclusiveMetricsEvaluatorWithTransforms as well.